### PR TITLE
[Windows] Add pcap service controller

### PR DESCRIPTION
**Pull request by gpotter2**
Previously, if the Npcap/Winpcap service was not started, scapy would not be working.

This PR:
- adds a detector + a util to ask the user if he wants to start the service automatically.
- makes scapy not crash when both powershell and cscript are unavailable
- restore VBS GUID change (I disabled it a quite long time ago, but it was indeed useful)

### CODE REVIEW

**Comment by p-l-** [comment]
That's very unusual in Scapy to clear the interpreter screen. Do you really need that? How bad would it be to let the user clear the window if he wants to?

**Comment by gpotter2** [comment]
@p-l- You're right, I do not need that in any way, and it might remove error messages

**Comment by gpotter2** [comment]
@p-l- That should be good

**Comment by gpotter2** [comment]
@guedou or @p-l- Could this be reviewed/merge ? It's quite important to the windows support and would be needed in https://github.com/secdev/scapy/pull/644

File: scapy/arch/windows/__init__.py [hunk 1]
```diff
@@ -410,10 +450,36 @@ def load_from_powershell(self):
...
-                                "You probably won't be able to send packets. "
-                                "Deactivating unneeded interfaces and restarting Scapy might help."
-                                "Check your winpcap and powershell installation, and access rights.", True)
+            _detect = pcap_service_status()
+            def _ask_user():
+                if not conf.interactive:
+                    return False
+                while True:
+                    _confir = raw_input("Do you want to start it ? (yes/no) [y]: ").lower().strip()
+                    if _confir in ["yes", "y", ""]:
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: While you are at it you might want to use _confir.lower() =)

File: scapy/arch/windows/__init__.py [hunk 2]
```diff
@@ -410,10 +450,36 @@ def load_from_powershell(self):
...
+                                                 or not hasattr(self, "restarted_adapter")):
+                warning("Scapy has detected that your pcap service is not running !")
+                if not conf.interactive or _ask_user():
+                    succeed = pcap_service_start(askadmin=conf.interactive)
+                    self.restarted_adapter = True
+                    if succeed:
+                        log_loading.info("Pcap service started !")
+                        self.load_from_powershell()
+                        return
+                warning("Could not start the pcap service ! "
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: You could probably define a variable to store the warning message.

File: scapy/arch/windows/__init__.py (hunk 1 above)
Comment (gpotter2) [inline scapy/arch/windows/__init__.py]: > _confir = raw_input("Do you want to start it ? (yes/no) [y]: ").lower().strip()
=)

File: scapy/arch/windows/__init__.py (hunk 2 above)
Comment (gpotter2) [inline scapy/arch/windows/__init__.py]: True

**Comment by gpotter2** [comment]
@guedou Hello, any news on this one ?
Thanks !

File: scapy/arch/windows/__init__.py [hunk 3]
```diff
@@ -95,24 +99,24 @@ def _vbs_exec_code(code, split_tag="@"):
...
+def _vbs_get_hardware_iface_guid(devid):
     try:
         devid = str(int(devid) + 1)
-        guid = _vbs_exec_code("""WScript.Echo CreateObject("WScript.Shell").RegRead("HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\NetworkCards\\%s\\ServiceName")
-""" % devid).__iter__().next()
+        guid = iter(_vbs_exec_code("""WScript.Echo CreateObject("WScript.Shell").RegRead("HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\NetworkCards\\%s\\ServiceName")
+""" % devid)).next()
         if guid.startswith('{') and guid.endswith('}\n'):
             return guid[:-1]
+        elif guid.startswith('{') and guid.endswith('}'):
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: Can't you merge the two guid blocs ?

File: scapy/arch/windows/__init__.py [hunk 4]
```diff
@@ -397,11 +400,48 @@ def is_invalid(self):
     def __repr__(self):
         return "<%s %s %s>" % (self.__class__.__name__, self.name, self.guid)
 
+def _pcap_service():
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: s/_pcap_service/pcap_service_name/

File: scapy/arch/windows/__init__.py [hunk 5]
```diff
@@ -397,11 +400,48 @@ def is_invalid(self):
...
+        return (None, None, None)
+    for i in exec_query(['Get-Service', 'npcap'], ['Name', 'DisplayName', 'Status']):
+        name = i[0]
+        description = i[1]
+        started = (not i[2].lower().strip() == 'stopped')
+        if name == _pcap_service():
+            return (name, description, started)
+    return (None, None, None)
+
+def _pcap_service_control(action, askadmin=True):
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: s/_pcap_service_control/pcap_service_control/

It seems better this way.

File: scapy/arch/windows/__init__.py [hunk 6]
```diff
@@ -561,7 +626,7 @@ def read_routes():
         else:
             routes = _read_routes_7()
     except Exception as e:    
-        warning("Error building scapy routing table : %s" % str(e), True)
+        warning("Error building scapy ipv4 routing table : %s" % str(e), True)
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: s/ipv4/IPv4/

File: scapy/arch/windows/__init__.py [hunk 7]
```diff
@@ -735,13 +800,15 @@ def _read_routes6_7():
...
+    if not conf.prog.os_access:
+        return routes6
     try:
         if is_new_release():
             routes6 = _read_routes6_post2008()
         else:
             routes6 = _read_routes6_7()
     except Exception as e:    
-        warning("Error building scapy routing table : %s" % str(e), True)
+        warning("Error building scapy ipv6 routing table : %s" % str(e), True)
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: s/ipv6/IPv6/

**Comment by guedou** [comment]
@gpotter2 I made some more comments. I am not a Windows person, so I can only trust you on the content of this PR =)

**Comment by gpotter2** [comment]
@guedou Thanks !
My work basically adds an util for npcap + fixes some VBS access on Windows 7

**Comment by guedou** [comment]
The `__init__.py` file is getting bigger and bigger. Will it make sense to move some parts of this PR to submodules lile `service.py` ?

File: scapy/arch/windows/__init__.py [hunk 8]
```diff
@@ -234,20 +236,23 @@ def win_find_exe(filename, installsubdir=None, env="ProgramFiles"):
...
-    # that must be in the path of the file
-    external_prog_list = {"AcroRd32" : "", "gsview32" : "", "dot" : "graph", "windump" : "", "tshark" : "",
-                          "tcpreplay" : "", "hexer" : "", "sox" : "", "wireshark" : ""}
     _default = "<System default>"
     def __init__(self):
-        _deep_lookup(self.external_prog_list)
+        # This is a dict containing the name of the .exe and a keyword
+        # that must be in the path of the file
+        external_prog_list = {"AcroRd32" : "", "gsview32" : "", "dot" : "graph", "windump" : "", "tshark" : "",
+                          "tcpreplay" : "", "hexer" : "", "sox" : "", "wireshark" : ""}
```
Comment (guedou) [inline scapy/arch/windows/__init__.py]: Could you align this correctly and shorten the lines length ?

**Comment by guedou** [comment]
I made a final comment related to alignment. Otherwise, it is good to be merged.

**Comment by gpotter2** [comment]
@guedou 
> Will it make sense to move some parts of this PR to submodules lile service.py ?

I think we can leave all the `powershell` access in a single file...

File: scapy/arch/windows/__init__.py (hunk 8 above)
Comment (gpotter2) [inline scapy/arch/windows/__init__.py]: Done

**Pull request merged**